torch/distributed/_functional_collectives.py

```diff
@@ -0,0 +1,196 @@
+            rankset = []
+            stride = -1
+            for rs in nested_list:
+                rankset.extend(rs)
+                if stride != -1 and stride != len(rs):
+                    raise ValueError(f"group sizes must be identical found {stride} and {len(rs)}")
+                stride = len(rs)
+        else:
+            rankset = cast(List[int], group)
+            stride = len(rankset)
+    elif isinstance(group, dist.ProcessGroup):
+        rankset = dist.get_process_group_ranks(group)
+        stride = len(rankset)
+        tag = tag or c10d._get_group_tag(group)
+    elif isinstance(group, dt.DeviceMesh):
+        rankset = group.mesh.flatten().tolist()
+        stride = len(rankset)
+        stride = group.mesh.size(0)
+        rankset = group.mesh.swapdims(-1, 0).reshape(-1, stride).flatten().tolist()
+        tag = tag or c10d._get_group_tag(group.get_dim_groups()[0])
+    elif isinstance(group, tuple):
+        if len(group) == 2 and isinstance(group[0], dt.DeviceMesh) and isinstance(group[1], int):
+            dmesh = group[0]
+            dim = group[1]
+            stride = dmesh.mesh.size(dim)
+            rankset = dmesh.mesh.swapdims(-1, dim).reshape(-1, stride).flatten().tolist()
+            tag = tag or c10d._get_group_tag(dmesh.get_dim_groups()[dim])
+        else:
+            raise ValueError("Invalid tuple for group must be (DeviceMesh, int)")
+    else:
+        raise ValueError("Invalid type for group, must be one of List, Processgroup, DeviceMesh or (DeviceMesh, int).")
+
+    return (tag, rankset, stride)
+
+
+class _WrapTensor(Function):
+    """
+    Autograd function to ensure autograd works through the wrapping
+    """
+    @staticmethod
+    def forward(ctx, tensor, reduceOp, stride):
+        ctx.reduceOp = reduceOp.upper()
+        ctx.stride = stride
+        return AsyncCollectiveTensor(tensor)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        if ctx.reduceOp == "SUM":
+            grad_output = grad_output / ctx.stride
+        return (grad_output, None, None)
+
+
+def all_reduce(self: torch.Tensor, reduceOp: str, group: RANK_TYPES, tag: str = ""):
+    """
+    Reduces the tensor data across all machines in such a way that all get
+    the final result.
+
+    The input tensor is left unmodified.
+
+    Group can be one of:
+        List[int]: ranks participating in the collective.
+        List[List[int]]: 2D mesh of ranks taking part of this collective in MPMD.
+        ProcessGroup: Will perform a collective using the ranks and tag of the PG.
+        DeviceMesh: Do a SPMD collective over all ranks of the mesh
+        (DeviceMesh, int): Do a MPMD collective over one dimension of the DeviceMesh
+
+    :: N.B. If you pass a PG or a 1D list to perform a MPMD collective, the compiler won't be able to recover
+    that information and perform collective algebraic optimization. Use other forms of input for that.
+    """
+    tag, rankset, stride = _expand_group(group, tag)
```

Comment:
I'm wondering if we could also lazily create subgroups when calling collective base on RANK_TYPES and tag? 

i.e. I would like to see DeviceMesh not creating those process group by hand, instead the functional collectives create and maintain those subgroups directly. This is beneficial for DTensor so that we wouldn't worry about pickling DTensor as `DeviceMesh` does not hold process groups.

Reply:
Yes, that's possible with the current design and I intend to do so once I modify DMesh.